app/src/lib/dragging/dropzone.ts

```diff
@@ -1,37 +1,123 @@
-export interface Dropzone {
+export interface DropzoneConfiguration {
 	disabled: boolean;
-	active: string;
-	hover: string;
 	accepts: (data: any) => boolean;
 	onDrop: (data: any) => Promise<void> | void;
+	onActivationStart: () => void;
+	onActivationEnd: () => void;
+	onHoverStart: () => void;
+	onHoverEnd: () => void;
+	target: string;
 }
+export class Dropzone {
+	private active: boolean = false;
 
-const defaultDropzoneOptions: Dropzone = {
-	disabled: false,
-	active: 'dropzone-active',
-	hover: 'dropzone-hover',
-	accepts: (data) => data === 'default',
-	onDrop: () => {}
-};
+	private registered: boolean = false;
+	// In order to propperly deregister functions we need to use the same
+	// reference so we must store the function after we bind the reference
+	private registeredOnDrop?: (e: DragEvent) => any;
+	private registeredOnDragEnter?: (e: DragEvent) => any;
+	private registeredOnDragLeave?: (e: DragEvent) => any;
 
-export function dropzone(node: HTMLElement, opts: Partial<Dropzone> | undefined) {
-	let currentOptions = { ...defaultDropzoneOptions, ...opts };
+	private target: HTMLElement;
 
-	function setup(opts: Partial<Dropzone> | undefined) {
-		currentOptions = { ...defaultDropzoneOptions, ...opts };
-		if (currentOptions.disabled) return;
+	constructor(
+		private configuration: DropzoneConfiguration,
+		node: HTMLElement
+	) {
+		const child = node.querySelector<HTMLElement>(configuration.target);
 
-		register(node, currentOptions);
+		if (child) {
+			this.target = child;
+		} else {
+			this.target = node;
+		}
+	}
+
+	async register(data: any) {
+		if (!this.configuration.accepts(await data)) return;
+		if (this.registered) {
+			this.unregister();
+		}
+
+		// Register listeners and keep references to the functions
+		this.registered = true;
+
+		this.registeredOnDrop = async (e: DragEvent) => await this.onDrop(e, data);
+		this.registeredOnDragEnter = this.onDragEnter.bind(this);
+		this.registeredOnDragLeave = this.onDragLeave.bind(this);
```

Comment:
I do make the assumption that the user can't drag two things at once, but that seems like a pretty sound assumption to me